FM_PC_ExternalPrice_Redeeming_v1_Test

scab24 opened the pull request:
Tests for the contract:

- [FM_PC_ExternalPrice_Redeeming_v1_Test.t.sol](https://github.com/ZealynxSecurity/InverterContracts/blob/G-dev-1-test/test/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1_Test.t.sol)
- [LM_ManualExternalPriceSetter_v1_Test.t.sol](https://github.com/ZealynxSecurity/InverterContracts/blob/G-dev-1-test/test/modules/fundingManager/oracle/LM_ManualExternalPriceSetter_v1_Test.t.sol)

FHieser committed:
Fix: Remove redeemed amount from balance check

* Remove unnecessary safety check

* Abstract projectFeeCollected

* Create tests

* Fmt

* Update src/modules/fundingManager/bondingCurve/interfaces/IRedeemingBondingCurveBase_v1.sol

Co-authored-by: Marvin <43185740+Zitzak@users.noreply.github.com>

* Add check to _handleCollateralTokensAfterSell

* Revert "Add check to _handleCollateralTokensAfterSell"

This reverts commit 32f9bb353784c05975e43c69d3006de8c79e8b58.

* Readd the check but only for projectCollateralFeeCollected

* Update IFM_BC_Bancor_Redeeming_VirtualSupply_v1.sol

* fmt

* Apply comment suggestions from code review

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

* Rename Error

* Apply suggestions from code review

* Fix: Resolve version mismatches

* Chore: Resolve formatting issue

---------

Co-authored-by: Marvin <43185740+Zitzak@users.noreply.github.com>
Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

vendrell46 committed:
Feat: FM_PC_ExternalPriceRedeeming, ManualExternalPriceSetter and ERC20Issuance_Blacklist

* Contract structure defined

* removed IFundingManager extension

* IFundingManager_v1 needed

* extended IRedeemingBondingCurveBase_v1 as well

* ExternalPriceSetter, blacklist, Oracle and interfaces added

* formatting files and adapting to the code of conduct

* ManualExternalPriceSetter_v1 modified as LM

* added fees checks, paymentOrder to the process and defined role for blacklist

* managing roles, renaming, and removing unneeded things

* whitelist logic removed, blacklist fixes

* missing logic implemented

* blacklist fixes and custom error added

* switched from role to mapping for blacklist manager

* moved blacklist files to external, modified sell, and more context added

* PAYMENT_QUEUE_ROLE added, _sellOrder overriden

* emit_timestamp(IssuancePriceSet,RedemptionPriceSet),NatSpec

* fix: remove collateral transfer & collateral check

- The function `handleCollateralTokensAfterSell` was still expecting to
transfer collateral, which would revert in the current implementation.
- The _sellOrder` function had a conditional check which would let
the contract revert given there will not be any collateral in the
contract

* style: run make fmt for formating

* small fixes and documentation added

* docs added for issuance and redemption setters

* Rename ERC20Issuance_blacklist_v1.sol to ERC20Issuance_Blacklist_v1.sol

* Rename IERC20Issuance_blacklist_v1.sol to IERC20Issuance_Blacklist_v1.sol

* fixed executeRedemptionQueue, docs and renaming

* formatting natspec

* feat: add treasury collateral withdraw

* Fix fmt

* fix: typo

* fix: fmt

* fix: remove oracle from init

* fixed natspec

* modifier moved

* feat: update openRedemptionAmount

* Update src/external/token/ERC20Issuance_Blacklist_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

* Update src/external/token/ERC20Issuance_Blacklist_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

* applied suggestions

* Update src/external/token/ERC20Issuance_Blacklist_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

* Update src/external/token/ERC20Issuance_Blacklist_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

* Update src/external/token/ERC20Issuance_Blacklist_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

* formatting

* Update src/modules/fundingManager/oracle/interfaces/IFM_PC_ExternalPrice_Redeeming_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

* formatting

* Update src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

* formatting

* Modifier buyingIsEnabled and sellingIsEnabled added

* formatting, updating code from parent and adding interface check

* Update src/external/token/ERC20Issuance_Blacklist_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

* Update src/external/token/ERC20Issuance_Blacklist_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

* Update src/modules/fundingManager/oracle/interfaces/ILM_ManualExternalPriceSetter_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

* Update src/modules/fundingManager/oracle/interfaces/ILM_ManualExternalPriceSetter_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

* Update src/modules/fundingManager/oracle/interfaces/IOraclePrice_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

* Update src/modules/fundingManager/oracle/interfaces/IOraclePrice_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

* Update src/external/token/interfaces/IERC20Issuance_Blacklist_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

* Update src/external/token/interfaces/IERC20Issuance_Blacklist_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

* Update src/external/token/interfaces/IERC20Issuance_Blacklist_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

* Update src/external/token/interfaces/IERC20Issuance_Blacklist_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

* Update src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

* Update src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

* Update src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

* Update src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

* Update src/modules/fundingManager/oracle/LM_ManualExternalPriceSetter_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

* Update src/modules/fundingManager/oracle/LM_ManualExternalPriceSetter_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

* formatting

* formatting

* moved constants

* changed notice to dev

* adding named returned variable

* modifier renamed and buyingIsEnabled added to buy

* return named variables to interface

* formatting

* fix: revert during deployment

* fix issue with authorization

* fix: interface check

* interface natspec same as contract

* line too long fixed

* removed dot from title

* _ensureTokenBalance overriden

* remove function

* fix fmt

* Update src/modules/fundingManager/oracle/interfaces/IOraclePrice_v1.sol

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

* fix: Move LM_ManualPriceSetter into LM folder

---------

Co-authored-by: scab24 <git.seco@protonmail.com>
Co-authored-by: Zitzak <marvin88gr@gmail.com>
Co-authored-by: Marvin <43185740+Zitzak@users.noreply.github.com>
Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>
Co-authored-by: Sergio <94926493+scab24@users.noreply.github.com>

Zitzak committed:
fix: decimal conversion error

* fix: decimal conversion error

* fix: add 10^tokenDecimal

Zitzak committed:
Feat: Add deduction of openRedemptionAmount

Co-authored-by: Marvin Kruse <marvinkruse@users.noreply.github.com>

scab24 committed:
FM_PC_ExternalPrice_Redeeming_v1_Test

scab24 committed:
Fix_Audit + Tests

Zitzak reviewed (changes requested):
Only reviewed implementation. Will be looking into the testing now

Zitzak commented on test/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1_Test.t.sol:
```diff
@@ -0,0 +1,2012 @@
+    "test/modules/fundingManager/bondingCurve/utils/mocks/RedeemingBondingCurveBaseV1Mock.sol";
+import {BancorFormula} from "@fm/bondingCurve/formulas/BancorFormula.sol";
+import {
+    PP_Streaming_v1,
```
Lots of unused imports ser

Zitzak commented on test/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1_Test.t.sol:
```diff
@@ -0,0 +1,2012 @@
+        _authorizer = authorizer;
+
+        //paymentProcessor
+        address PaymentImpl = address(new PP_Streaming_v1AccessMock());
```
Not sure why the default PP that is deployed in the `ModuleTest` isn't used?

Zitzak reviewed (changes requested):
I'm finding a lot of things setup wise which should be changed. Part of this is probably also on me as I could have been involved earlier.

I will be stopping the review for today but continue tomorrow. Also I'm updating most of the points I'm noticing on a branch at the same time while reviewing as it is probably a lot easier for me to go in and change it right away.

Will keep you updated!

Zitzak committed:
refactor: remove duplicate ERC20Mock, extend existing with decimals

Zitzak committed:
refactor: initial cleanup LM_ManualExternalPriceSetter_v1

- removed unused imports
- moved files to right directory
- renamed files
- cleaned up setup by removing unnecessary initialization, instead
reuse existing setup from ModuleTest
- format code (make fmt)

Zitzak committed:
refactor: initial setup ERC20Issuance_Blacklist_v1
- Moved file
- Removed unnecessary imports
- Use only exposed mock for testing

Zitzak committed:
refactor: FM_PC_ExternalPrice_Redeeming_v1_Test

- Remove imports
- Refactor most of the setup
- Add exposed mock
- Removed OraclePriceMock and InvalidMock
- Added simpler oracle mockf
- fmt

Zitzak committed:
refactor: remove import and fmt of LM_ManualExternalPriceSetter test

Zitzak commented on test/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1_Test.t.sol:
```diff
@@ -0,0 +1,2012 @@
+    // Setup
+    // ═══════════════════════════════════════════════════════════════════════════════════════════════════════
+
+    function setUp() public {
```
Most of the setup should be done differently and has been implemented in the PR I opened

leeftk commented on test/modules/fundingManager/oracle/LM_ManualExternalPriceSetter_v1_Test.t.sol:
```diff
@@ -0,0 +1,548 @@
+        vm.startPrank(admin);
+
+        // Create mock tokens with different decimals
+        inputToken = new MockERC20(6); // Like USDC
```
This may be a bit of a naive question on my part so apologies but Im not seeing where we can adjust the decimals for the ERC20Mock that is already provided for us, could this be the reason that @scab24 decided to create a new contract? In this case this ERC20 has 6 decimals

leeftk replied:
I saw your implementation of this on your PR please disregard this comment

Zitzak replied:
Yes you are right, and I noticed that we didn't have the ability to set decimals. I indeed ended up implementing it as defined in this [comment](https://github.com/InverterNetwork/contracts/pull/706#discussion_r1907657810)

leeftk commented on test/modules/fundingManager/oracle/utils/mocks/OraclePriceMock.sol:
```diff
@@ -0,0 +1,41 @@
+import "@lm/interfaces/IOraclePrice_v1.sol";
+import "@oz/utils/introspection/ERC165.sol";
+
+contract OraclePriceMock is IOraclePrice_v1, ERC165 {
```
Yes I removed this in the latest PR

leeftk commented on src/modules/logicModule/LM_ManualExternalPriceSetter_v1.sol:
```diff
@@ -143,7 +140,25 @@ contract LM_ManualExternalPriceSetter_v1 is
+    function setIssuanceAndRedemptionPrice(
+        uint issuancePrice_,
+        uint redemptionPrice_
+    ) external onlyModuleRole(PRICE_SETTER_ROLE) {
```
Made this change in my latest PR

leeftk commented on src/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1.sol:
```diff
@@ -578,26 +570,12 @@ contract FM_PC_ExternalPrice_Redeeming_v1 is
-
         // Create and emit the order.
         _createAndEmitOrder(
             _receiver, _depositAmount, collateralRedeemAmount, issuanceFeeAmount
```
Made this change in my latest PR

leeftk replied:
Addressed in the feature branch

Zitzak committed:
Merge pull request #12 from ZealynxSecurity/updateTesting

Refactor: Initial refactor of testing

Zitzak commented on test/modules/fundingManager/oracle/FM_PC_ExternalPrice_Redeeming_v1_Test.t.sol:
```diff
@@ -0,0 +1,2012 @@
+ * @notice Test contract for FM_PC_ExternalPrice_Redeeming_v1
+ */
+contract FM_PC_ExternalPrice_Redeeming_v1_Test is ModuleTest {
+
```
Leaving this one open as a reminder to do the internal function testing. The access mock got added through my PR

Zitzak committed:
fix: compile error

scab24 committed:
Fix=> testExternalBuy_revertsGivenNonWhitelistedUser, testExternalBuy_succeedsGivenMaxAmount and comment testExternalQueue_succeedsGivenSingleRedemptionOrder

Zitzak commented:
Closing this PR, as all the changes have been merged into [ZealynxSecurity/Feature/FM_PC_ExternalPrice](https://github.com/ZealynxSecurity/InverterContracts/tree/ZealynxSecurity/Feature/FM_PC_ExternalPrice) branch
@vendrell46 @scab24 @leeftk
